armory/instrument/instrument.py

```diff
@@ -0,0 +1,690 @@
+    def is_measuring(self, probe_variable):
+        return bool(
+            self.mapper.map_probe_update_to_meter_input(
+                probe_variable, self.context["stage"]
+            )
+        )
+
+    def update(self, probe_variable, value):
+        meters_args = self.mapper.map_probe_update_to_meter_input(
+            probe_variable, self.context["stage"]
+        )
+        if not meters_args:
+            raise ValueError("No meters are measuring")
+        for meter, arg in meters_args:
+            meter.set(arg, value, self.context["batch"])
+
+    def connect_meter(self, meter, use_default_writers=True):
+        """
+        Connect meter. If use_default_writers, connect to all default writers as well
+        """
+        if use_default_writers:
+            for writer in self.default_writers:
+                meter.add_writer(writer)
+
+        if meter in self.meters:
+            return
+
+        self.meters.append(meter)
+        self.mapper.connect_meter(meter)
+
+    def connect_writer(self, writer, meters=None, default=False):
+        """
+        Convenience method to add writer to all (or a subset of meters)
+
+        meters - if None, add to all current meters
+            otherwise, meters should be a list of names and/or Meter objects
+                the writer is connected to those meters
+            if meters is an empty list, the writer will be added but no meters connected
+
+        default - if True, writer is automatically added to each new meter added
+        """
+        if meters is None:
+            meters_list = self.meters
+        else:
+            # parse meters list
+            meters_list = []
+            for m in meters:
+                if isinstance(m, str):
+                    pass
+                elif isinstance(m, Meter):
+                    if m in self.meters:
+                        meters_list.append(m)
+                else:
+                    raise ValueError(
+                        f"meters includes {m}, which is neither a str nor Meter"
+                    )
+
+        for meter in meters_list:
+            meter.add_writer(writer)
+
+        if default and writer not in self.default_writers:
+            self.default_writers.append(writer)
+        if writer not in self.writers:
+            self.writers.append(writer)
+
+    def record(self, name, result):
+        """
+        Push a record to the default writers
+        """
+        if not self.default_writers:
+            log.warning(f"No default writers to record {name}:{result} to")
+        for writer in self.default_writers:
+            writer.write((name, self.context["batch"], result))
+
+    def close(self):
+        if self.closed:
+            return
+
+        for meter in self.meters:
+            meter.finalize()
+
+        for writer in self.writers:
+            writer.close()
+
+        self.closed = True
+
+
+class Meter:
+    def __init__(
+        self,
+        name,
+        metric,
```

Comment:
could it make sense to allow metric to be `None`, in which case it just records whatever value you're measuring (in other words, basically an identity function). I'm again imagining the use case, which I think would be common, of users just wanting to store an intermediate model output

Reply:
perhaps I could use achieve this by passing in a simple metric function that merely returns the input, but that feels a little hacky

Reply:
The main issue is when you have multiple values and multiple batches - you need to do some weird zipping and unzipping to make this work properly.